diff -Naur linux-5.1_clean/sound/soc/ti/davinci-mcasp.c linux-5.1/sound/soc/ti/davinci-mcasp.c
--- linux-5.1_clean/sound/soc/ti/davinci-mcasp.c	2019-05-06 02:42:58.000000000 +0200
+++ linux-5.1/sound/soc/ti/davinci-mcasp.c	2019-05-25 11:10:23.751017143 +0200
@@ -845,16 +845,16 @@
 					rx_ser < max_active_serializers) {
 			clear_bit(PIN_BIT_AXR(i), &mcasp->pdir);
 			rx_ser++;
-		} else if (mcasp->serial_dir[i] == INACTIVE_MODE) {
+		} else /*if (mcasp->serial_dir[i] == INACTIVE_MODE)*/ {
 			mcasp_mod_bits(mcasp, DAVINCI_MCASP_XRSRCTL_REG(i),
 				       SRMOD_INACTIVE, SRMOD_MASK);
 			clear_bit(PIN_BIT_AXR(i), &mcasp->pdir);
-		} else if (mcasp->serial_dir[i] == TX_MODE) {
+		} /*else if (mcasp->serial_dir[i] == TX_MODE) {
 			/* Unused TX pins, clear PDIR  */
-			mcasp_mod_bits(mcasp, DAVINCI_MCASP_XRSRCTL_REG(i),
+		/*	mcasp_mod_bits(mcasp, DAVINCI_MCASP_XRSRCTL_REG(i),
 				       mcasp->dismod, DISMOD_MASK);
 			clear_bit(PIN_BIT_AXR(i), &mcasp->pdir);
-		}
+		}*/
 	}
 
 	if (stream == SNDRV_PCM_STREAM_PLAYBACK) {
