diff -Naur linux-am33x-clean/src/linux-4.19/sound/soc/davinci/davinci-mcasp.c linux-am33x/src/linux-4.19/sound/soc/davinci/davinci-mcasp.c
--- linux-am33x-clean/src/linux-4.19/sound/soc/davinci/davinci-mcasp.c	2018-10-22 08:37:37.000000000 +0200
+++ linux-am33x/src/linux-4.19/sound/soc/davinci/davinci-mcasp.c	2018-12-30 10:59:58.256972686 +0100
@@ -781,7 +781,7 @@
 					rx_ser < max_active_serializers) {
 			mcasp_clr_bits(mcasp, DAVINCI_MCASP_PDIR_REG, AXR(i));
 			rx_ser++;
-		} else if (mcasp->serial_dir[i] == INACTIVE_MODE) {
+		} else {
 			mcasp_mod_bits(mcasp, DAVINCI_MCASP_XRSRCTL_REG(i),
 				       SRMOD_INACTIVE, SRMOD_MASK);
 		}
